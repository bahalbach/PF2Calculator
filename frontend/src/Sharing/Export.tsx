import { Button, Paper, TextField } from "@mui/material";
import Grid from "@mui/material/Grid";
import React, { useEffect, useState } from "react";

import { useAppDispatch, useAppSelector } from "../App/hooks";
// import { exampleRoutines } from "../Model/exampleRoutines";
import { importStates } from "../Model/types";
import {
  importRoutine,
  selectImportState,
  selectSelectedRoutineObject,
} from "../Routines/RoutineSlice/routineSlice";
// import { useSelector } from "react-redux";

/* TODO: 
  Convert to mui
  Add save routines button
  Add load example routines button
*/
const ImportExport = () => {
  const routine = useAppSelector(selectSelectedRoutineObject);
  const importState = useAppSelector(selectImportState);
  const [messageSeen, setMessageSeen] = useState(false);
  const [textValue, setTextValue] = useState("");

  const dispatch = useAppDispatch();

  useEffect(() => {
    if (!messageSeen) {
      if (importState === importStates.Failure) {
        setTextValue("Import failed");
        setMessageSeen(true);
      } else if (importState === importStates.Successful) {
        setTextValue("Import successful");
        setMessageSeen(true);
      }
    }
  }, [importState, messageSeen]);

  const tryToAddRoutine = () => {
    dispatch(importRoutine(textValue));
    setMessageSeen(false);
  };
  const paste = () => {
    setTextValue(JSON.stringify(routine));
  };

  //   const saveRoutines = () => {try {
  //       const serialState = JSON.stringify(state);
  //       localStorage.setItem('appState', serialState);
  //     } catch(err) {
  //         console.log(err);
  //     }
  // };
  // const loadSavedRoutines = () => {};
  // const addExampleRoutines = () => {
  //   for (let r of exampleRoutines) {
  //     dispatch(importRoutine(r));
  //   }
  // };
  const copy = () => {
    if (navigator.clipboard !== undefined) {
      // textAreaRef.current.select();
      // textAreaRef!.current.setSelectionRange(0, 99999); /* For mobile devices */
      navigator.clipboard.writeText(textValue);
      setTextValue("Copied");
      // document.execCommand("copy");
    }
  };
  return (
    <Paper sx={{ p: 1 }}>
      <Grid container>
        <Grid>
          <Button onClick={tryToAddRoutine}>Import</Button>
        </Grid>
        <Grid>
          <Button onClick={paste}>Export</Button>
        </Grid>
        <Grid>
          <Button onClick={copy}>Copy</Button>
        </Grid>
        {/* <Grid item>
          <Button onClick={saveRoutines}>Save Routines</Button>
        </Grid> */}
        {/* <Grid item>
          <Button onClick={loadSavedRoutines}>Load Saved Routines</Button>
        </Grid> */}
        {/* <Grid item>
          <Button onClick={addExampleRoutines}>Add example Routines</Button>
        </Grid> */}
        <Grid size={12}>
          <TextField
            fullWidth
            multiline
            maxRows={5}
            label="Routine code"
            placeholder="Enter routine code here to import."
            value={textValue}
            onChange={(e) => setTextValue(e.target.value)}
          />
        </Grid>
      </Grid>
    </Paper>
  );
};

export default ImportExport;
