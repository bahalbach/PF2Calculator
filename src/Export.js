import React from "react";
// import { useSelector } from "react-redux";

const ImportExport = () => {
  return <div className="box">Import and Export Here</div>;
};

export default ImportExport;
