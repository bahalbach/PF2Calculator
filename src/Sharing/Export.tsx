import { Button, Grid, Paper, TextField, Typography } from "@mui/material";
import React, { useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import { exampleRoutines } from "../Model/exampleRoutines";
import {
  importRoutine,
  selectSelectedRoutineObject,
} from "../Routines/routineSlice";
// import { useSelector } from "react-redux";

/* TODO: 
  Convert to mui
  Add save routines button
  Add load example routines button
*/
const ImportExport = () => {
  const routine = useSelector(selectSelectedRoutineObject);
  // const textAreaRef: RefObject<HTMLTextAreaElement> = useRef(null);
  const [textValue, setTextValue] = useState("");
  const dispatch = useDispatch();

  const tryToAddRoutine = () => {
    dispatch(importRoutine(textValue));
  };
  const paste = () => {
    setTextValue(JSON.stringify(routine));
  };

  //   const saveRoutines = () => {try {
  //       const serialState = JSON.stringify(state);
  //       localStorage.setItem('appState', serialState);
  //     } catch(err) {
  //         console.log(err);
  //     }
  // };
  // const loadSavedRoutines = () => {};
  const addExampleRoutines = () => {
    for (let r of exampleRoutines) {
      dispatch(importRoutine(r));
    }
  };
  // const copy = () => {
  //   if (textAreaRef.current) {
  //     textAreaRef.current.select();
  //     textAreaRef!.current.setSelectionRange(0, 99999); /* For mobile devices */
  //     navigator.clipboard.writeText(textAreaRef.current.value);
  //     // document.execCommand("copy");
  //   }
  // };
  return (
    <Paper sx={{ p: 1 }}>
      <Typography variant="h5">Import and Export Here:</Typography>
      <Grid container>
        <Grid item>
          <Button onClick={tryToAddRoutine}>Import</Button>
        </Grid>
        <Grid item>
          <Button onClick={paste}>Export</Button>
        </Grid>
        {/* <Grid item>
          <Button onClick={saveRoutines}>Save Routines</Button>
        </Grid> */}
        {/* <Grid item>
          <Button onClick={loadSavedRoutines}>Load Saved Routines</Button>
        </Grid> */}
        <Grid item>
          <Button onClick={addExampleRoutines}>Add example Routines</Button>
        </Grid>
        <Grid item xs={12}>
          <TextField
            fullWidth
            multiline
            label="Routine code"
            placeholder="Enter routine code here to import."
            value={textValue}
            onChange={(e) => setTextValue(e.target.value)}
          />
        </Grid>
      </Grid>
    </Paper>
  );
};

export default ImportExport;
