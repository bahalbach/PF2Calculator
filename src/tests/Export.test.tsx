import React from "react";
import { render } from "./test-utils";

import Export from "../Sharing/Export";
import { fireEvent, screen } from "@testing-library/dom";

describe("Export", () => {
  test("renders correctly", () => {
    const { asFragment } = render(<Export />);
    expect(asFragment()).toMatchSnapshot();
  });

  test("Export and Import functions work correctly", () => {
    render(<Export />);
    expect(screen.getByRole("textbox")).toHaveValue("Import successful");

    fireEvent.click(screen.getByText("Export"));
    expect(screen.getByRole("textbox")).not.toHaveValue("Import successful");

    fireEvent.click(screen.getByText("Import"));
    expect(screen.getByRole("textbox")).toHaveValue("Import successful");

    fireEvent.click(screen.getByText("Import"));
    expect(screen.getByRole("textbox")).toHaveValue("Import failed");
  });

  // Can't test 'Copy' because no navigator.clipboard
});
