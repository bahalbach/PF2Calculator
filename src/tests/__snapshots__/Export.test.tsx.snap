// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Export renders correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-3ze8r2-MuiPaper-root"
  >
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-undefined css-11lq3yg-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      >
        <button
          class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-1ju4386-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Import
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      >
        <button
          class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-1ju4386-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Export
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      >
        <button
          class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-1ju4386-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Copy
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1em9ut6-MuiGrid-root"
      >
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary MuiFormLabel-filled css-m9off1-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="true"
          >
            Routine code
          </label>
          <div
            class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-multiline css-1co3et4-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <textarea
              aria-invalid="false"
              class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
              placeholder="Enter routine code here to import."
              style="height: 0px; overflow: hidden;"
            >
              Import successful
            </textarea>
            <textarea
              aria-hidden="true"
              class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
              readonly=""
              style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding: 0px; width: 100%;"
              tabindex="-1"
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-1z7n62"
              >
                <span>
                  Routine code
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
